### PR TITLE
Subscription support

### PR DESCRIPTION
**Pull request by klaseskilson**
This adds support for both the App Store's and Play Store's subscriptions. By using the verifiers' `verify_subscription` methods, you can verify subscriptions much like you can verify subscriptions much like you can verify single purchases.

Hit me with your feedback! I'd love to hear your thoughts on what I've done. 👍

### CODE REVIEW

File: lib/candy_check/utils/attribute_reader.rb [hunk 1]
```diff
@@ -18,6 +18,14 @@ def read_integer(field)
         (val = read(field)) && val.to_i
       end
 
+      # @return [bool] if value is either 'true' or 'false'
+      # @return [nil] if value is not 'true'/'false'
+      def read_bool(field)
+        val = read(field)
+        nil unless %w(false true).include?(val)
```
Comment (splattael) [inline lib/candy_check/utils/attribute_reader.rb]: Is a possible `return` Is missing?

Comment (klaseskilson) [inline lib/candy_check/utils/attribute_reader.rb]: You are totally correct! I thought I fixed that.

File: lib/candy_check/app_store/verifier.rb [hunk 2]
```diff
@@ -30,6 +30,18 @@ def initialize(config)
       # @return [Receipt] if successful
       # @return [VerificationFailure] otherwise
       def verify(receipt_data, secret = nil)
+        @verifier = Verification
+        fetch_receipt_information(receipt_data, secret)
+      end
+
+      def verify_subscription(receipt_data, secret = nil)
```
Comment (jnbt) [inline lib/candy_check/app_store/verifier.rb]: It's a bit picky, but can you please add some documentation here.

File: lib/candy_check/play_store/subscription.rb [hunk 3]
```diff
@@ -0,0 +1,119 @@
+module CandyCheck
+  module PlayStore
+    # Describes a succeful subscription validation
+    class Subscription
+      include Utils::AttributeReader
+
+      # @return [Hash] the raw attributes returned from the server
+      attr_reader :attributes
+
+      # Values of paymentState
```
Comment (jnbt) [inline lib/candy_check/play_store/subscription.rb]: These docs need some adjustments, like: 
```ruby
# The payment of the subscription is pending (paymentState)
PAYMENT_PENDING = 0
# The payment of the subscription is received (paymentState)
PAYMENT_RECEIVED = 1
# The payment was canceled (cancelReason)
PAYMENT_CANCELED = 0
# The payment did fail during processing (cancelReason)
PAYMENT_FAILED = 1
```

File: lib/candy_check/play_store/subscription.rb [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        cancel_reason == PAYMENT_CANCELED
+      end
+
+      # Get number of overdue days. If this is negative, it is not overdue.
+      # @return [Integer]
+      def overdue_days
+        (Date.today - expires_at.to_date).to_i
+      end
+
+      def auto_renewing?
```
Comment (jnbt) [inline lib/candy_check/play_store/subscription.rb]: Could you add some docs to this attribute getter (and the following) as well?

File: spec/app_store/verifier_spec.rb [hunk 5]
```diff
@@ -80,6 +81,27 @@
...
+
+    it 'only uses production endpoint for normal failures' do
+      failure = get_failure(21_000)
+      with_mocked_verifier(failure) do
+        subject.verify_subscription(data, secret).must_be_same_as failure
+        assert_recorded([production_endpoint, data, secret])
+      end
+    end
+  end
+
```
Comment (jnbt) [inline spec/app_store/verifier_spec.rb]: Could you add a test for the case the wrong endpoint is hit, and the opposite endpoint must be retried? Just like `"retries production endpoint for redirect error"`

**Comment by jnbt** [comment]
@klaseskilson Big 👏 ! This is really an awesome contribution. I'm happy to merge it soon. Especially as I haven't used the subscriptions yet!

Could you have a look at some of my annotations? 

And another question: Do you have an Idea how to test it against the real APIs? I think the subscription verification is still lacking CLI support `lib/cli/app.rb`, but I'm more then willing to implement this myself. 
I think to test iOS we need to find / generate some receipt data for a subscription.
For Android a valid subscription?

**Comment by klaseskilson** [comment]
@jnbt Sounds great, I'll take a look at your comments asap. I left out CLI support for now, yes. If you don't mind implementing it yourself, go ahead. 👍

File: spec/app_store/verifier_spec.rb (hunk 5 above)
Comment (klaseskilson) [inline spec/app_store/verifier_spec.rb]: Sure, adding it! Sandbox accounts are definitely useful for subscriptions too. 👍

Comment (jnbt) [inline spec/app_store/verifier_spec.rb]: So verifying subscriptions works basically like standard purchases for iOS? (Except the part around the shared secret)

**Comment by klaseskilson** [comment]
@jnbt Pushed some changes now.

Regarding testing against the real APIs, I'm not sure either. I guess you would need some sort of real data or subscription info and the corresponding authorization keys. That might be semi-tricky to get a hold of, since it might contain sensitive information.

I guess some information could be stored in a CI environment to allow for tests that way, but that will create a tricky situation when it comes to local testing. Also in a CI situation, storing the P12 file securely is kind of tricky and keeping auth keys to Google up to date is a hassle.

**Comment by jnbt** [comment]
The same holds for all PlayStore verifications as they also require a secret. At least I did some local testing on my machine for these cases.

I think I try to setup an iOS and PlayStore test app and configure at least one product and one subscription. We might even use Travis to test it, as we could encrypt these data.

**Comment by jnbt** [comment]
Addition: For obvious reasons we could only test against sandbox... (but this is more than nothing.)

**Comment by klaseskilson** [comment]
Oh right, Travis supports encryption!

Is this something you want me to look at before merging this PR?

**Comment by jnbt** [comment]
I'll merge it in the current state. Thanks for all the work. Do you have an urgent use case? Otherwise I'd prefer to do the CLI part before publishing a new gem version. This would also allow some manual integration testing. Have you performed any real interactions against the subscription API while developing the extension?

**Comment by klaseskilson** [comment]
I tottaly agree that a way to test subscriptions manually should be made possible before publishing a new version. I'm fine waiting, I can pull it through git meanwhile.

Awesome work on the gem, thanks for merging! 👌

**Comment by jnbt** [comment]
I'll try to have a look at CLI this or next week.

**Comment by klaseskilson** [comment]
Awesome, let me know if you need any help regarding subscriptions!

**Comment by jnbt** [comment]
Thanks, I've already invited you to be a collaborator.

**Comment by klaseskilson** [comment]
Thank you! Looking forward to contributing with subscriptions! 🍻

**Pull request merged**